fields/src/macros.rs

```diff
@@ -144,25 +144,28 @@ macro_rules! sqrt_impl {
                     l_s.iter().take(j).sum::<u64>() + 1 + l_s.iter().skip(i + 1).sum::<u64>()
                 };
 
-                let calc_gamma = |i: usize, q_s: &[Vec<bool>], last: bool| -> $Self {
+                let calc_gamma = |i: usize, q_s: &[u64], last: bool| -> $Self {
                     let mut gamma = $Self::one();
                     if i != 0 {
-                        q_s.iter().zip(l_s.iter()).enumerate().for_each(|(j, (q_bits, l))| {
+                        q_s.iter().zip(l_s.iter()).enumerate().for_each(|(j, (q, l))| {
                             let mut kappa = calc_kappa(i, j, &l_s);
                             if last {
                                 kappa -= 1;
                             }
-                            q_bits.iter().enumerate().take(*l as usize).for_each(|(k, bit)| {
-                                if *bit {
+                            let mut value = *q;
+                            (0..*l as usize).for_each(|k| {
+                                let bit = value & 1 == 1;
+                                if bit {
                                     gamma *= $Self($P::POWERS_OF_ROOTS_OF_UNITY[(kappa as usize) + k], PhantomData);
                                 }
+                                value = value.wrapping_shr(1u32);
```

Comment:
this is a non-allocating conversion that's identical to the one in the `ToBits` impl for `u64`